src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/services/LinkedStudyServiceImpl.java

```diff
@@ -0,0 +1,160 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package uk.ac.ebi.ampt2d.metadata.persistence.services;
+
+import com.querydsl.core.types.Predicate;
+import org.springframework.beans.factory.annotation.Autowired;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.AccessionVersionEntityId;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.LinkedStudy;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.LinkedStudyId;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.QLinkedStudy;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.QStudy;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.Study;
+import uk.ac.ebi.ampt2d.metadata.persistence.repositories.LinkedStudyRepository;
+import uk.ac.ebi.ampt2d.metadata.persistence.repositories.StudyRepository;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+
+public class LinkedStudyServiceImpl implements LinkedStudyService {
+
+    @Autowired
+    private StudyRepository studyRepository;
+
+    @Autowired
+    private LinkedStudyRepository linkedStudyRepository;
+
+    private List<LinkedStudy> findLinkedStudies(AccessionVersionEntityId id) {
+        QLinkedStudy linkedStudy = QLinkedStudy.linkedStudy1;
+        Predicate predicate = linkedStudy.study.id.eq(id).or(linkedStudy.linkedStudy.id.eq(id));
+
+        return (List<LinkedStudy>) linkedStudyRepository.findAll(predicate);
+    }
+
+    private List<AccessionVersionEntityId> findLinkedStudyIds(AccessionVersionEntityId id) {
+        Set<AccessionVersionEntityId> ids = new HashSet<>();
+        Set<AccessionVersionEntityId> seen = new HashSet<>();
+        Stack<AccessionVersionEntityId> stack = new Stack<>();
+
+        seen.add(id);
+        stack.push(id);
+
+        while ( !stack.empty() ) {
+            AccessionVersionEntityId top = stack.pop();
+            List<LinkedStudy> linkedStudies = findLinkedStudies(top);
+
+            for ( LinkedStudy linkedStudy : linkedStudies ) {
+                Arrays.asList(linkedStudy.getStudy().getId(), linkedStudy.getLinkedStudy().getId())
+                        .forEach(studyId->{
+                            if ( !top.equals(studyId) && !seen.contains(studyId) ) {
+                                ids.add(studyId);
+                                seen.add(studyId);
+                                stack.push(studyId);
+                            }
+                        });
+            }
+        }
+
+        return ids.stream().collect(Collectors.toList());
+    }
+
+    @Override
+    public LinkedStudy findOne(AccessionVersionEntityId studyId, AccessionVersionEntityId linkedStudyId) {
```

Comment:
You could directly use the find by Id with the two possible ids.

Reply:
Fixed.